### DISCUSSION TITLE
Questions about programming for tearout/tearin

### DISCUSSION THREAD

**Discussion by vefatica** [microsoft/terminal#18271] [General]
I've got the TCC shell, which does a number of GUI things, working pretty well as long as its CASCADIA_HOSTING_WINDOW_CLASS window doesn't change.  Tearout/tearin poses new problems.  The answer to one question might let me handle this.  If not, I'll be back.  So here's the question.

Upon tearout/tearin, does the shell get a new PseudoConsoleWindow or does the old PseudoConsoleWindow get a new parent?

Any other insight would be appreciated.

**Comment by vefatica** [comment]
OK, that was a bad question.  I see now that, after tearout, neither the PseudoConsoleWindow nor its parent (the Cascadia window) changes.  That's exactly my problem.  I want the Cascadia window that my shell is actually _in_ (after tearout), not the one it used to be in.  How do I do that?

**Reply by lhecker** [reply]
The `GA_ROOTWINDOW` code that was suggested in the other discussion should theoretically work for this. If not, we have a bug in our code.

**Comment by vefatica** [comment]
I'm using that now.  Inside my `GetCascadiaWindow()` I have `HWND hWndCascadia = GetAncestor(GetConsoleWindow(), GA_ROOTOWNER);`.  And I use (my own) `PrintWindowInfo(GetConsoleWindow())`; and `PrintWindowInfo(GetCascadiaWindow())`;

Here's what I see, first, before tearout.

![image](https://github.com/user-attachments/assets/6f16982a-48ab-45a1-9f93-ae68603b7e8f)

Then after tearing out the 5948 tab; nothing has changed.

![image](https://github.com/user-attachments/assets/c6e608ce-0f84-4e65-bdff-7496026356f7)

Are you supposed to be giving the PseudoConsoleWindow a new parent?

**Reply by lhecker** [reply]
My understanding was that the root parent should be the WT window. I'm only starting to get into our windowing code though, so I may be wrong. I'll check it out.

**Reply by zadjii-msft** [reply]
I'm guessing I forgot to call SetParent or SetOwner or whatever in TermControl if the tab dragged out. That's my gut

**Reply by lhecker** [reply]
It actually works just fine for me, in all 3 versions of Windows Terminal. I'm using this code to test it:
```cpp
#define NOMINMAX
#define WIN32_LEAN_AND_MEAN
#include <Windows.h>
#include <cstdio>

int main() {
    HWND last_console = nullptr;
    HWND last_root = nullptr;

    for (;;) {
        const auto console = GetConsoleWindow();
        const auto root = GetAncestor(console, GA_ROOTOWNER);

        if (last_console != console || last_root != root) {
            last_console = console;
            last_root = root;
            printf("Console: %p, Root: %p\n", console, root);
        }

        Sleep(100);
    }
}
```

**Reply by DHowett** [reply]
Hey @vefatica, can you add timestamps (one second resolution is probably fine) to your repro and grab new screenshots before and after tearout?

**Reply by vefatica** [reply]
Sure, but I don't know what you mean ... time of day when I do the tests?  The two tests are several seconds apart.

**Reply by vefatica** [reply]
Now I'm using this.

![image](https://github.com/user-attachments/assets/f266b265-cfad-4a16-a0c0-27f51834d8ae)

Before:

![image](https://github.com/user-attachments/assets/6bb81228-bb19-406d-bd2f-064e122e55e7)

After:

![image](https://github.com/user-attachments/assets/2ff0dda0-7a72-48fb-84b1-5b263699ef2d)

**Reply by DHowett** [reply]
Can you print, on the screen, in string form, the literal actual time at which you run WindowCheck.exe?

Put another way, can you put the literal actual time in string form in the output of `PrintWindowInfo`?

**Reply by vefatica** [reply]
OK (give me a few minutes) ... but why?

**Reply by vefatica** [reply]
Is this good enough?

Before:

![image](https://github.com/user-attachments/assets/a2eea95a-c27c-4cc9-83fd-734a66de5eed)

After:

![image](https://github.com/user-attachments/assets/8432cc02-35d6-4abe-a11b-123a9067c32d)

**Reply by DHowett** [reply]
Yes, this is helpful. Thank you! I had had a theory, but it’s shot.

**Reply by vefatica** [reply]
OK, my fault (as is often the case).  I was making the console window message-only (with SetParent).  With that removed, I see the change of the root window.

Is that right ... SetParent(hWnd, HWND_MESSAGE) screws up future use of SetParent?  Is there a way to undo message-only?

We discussed the pseudo console window becoming visible a long time ago.  Have you taken steps to prevent that?

**Reply by DHowett** [reply]
Unfortunately, we never proved conclusively that it was actually the pseudoconsole window. The current theory is that it's the XAML input site or XAML CoreWindow (both of which are an implementation detail of the UI framework we are using in Terminal).

Leonard fixed a lot of Pseudoconsole hosting window issues in 1.21 and 1.22 regardless. It may be worth trying again to not reparent it.

**Reply by vefatica** [reply]
> It may be worth trying again to not reparent it.

Yeah, I'm doing that now and, so far so good.  I have an OpenConsole-specific EVENT_SYSTEM_FOREGROUND WinEvent hook that foregrounds the Cascadia window if anyone tries to foreground the console window (as TCC does).  Maybe that's doing the job.  Or maybe that's not necessary given Leonard's improvements.

I'm pretty sure it was the pseudoconsole window I was seeing.  I make it follow the Cascadia window around in size and position and when it shows (or used to show, fingers crossed) it's pretty obvious that that's what it is.

**Reply by vefatica** [reply]
> Leonard fixed a lot of Pseudoconsole hosting window issues in 1.21 and 1.22

> I have an OpenConsole-specific EVENT_SYSTEM_FOREGROUND WinEvent hook that foregrounds the Cascadia window if anyone tries to foreground the console window (as TCC does)

When I remove that hook I see the same thing I'd see if it were still there.  That is, in a situation where TCC tries to foreground the console window (when the user has dismissed a dialog), I see it become the foreground window only very momentarily (~ 2ms).  Below, I'm monitoring as fast as I can with command line tools.

```
11:05:43.195023 : "#32770"
11:05:43.195359 : "#32770"
11:05:43.195923 : "PseudoConsoleWindow"
11:05:43.196412 : "PseudoConsoleWindow"
11:05:43.196810 : "PseudoConsoleWindow"
11:05:43.197088 : "PseudoConsoleWindow"
11:05:43.197393 : "CASCADIA_HOSTING_WINDOW_CLASS"
11:05:43.197773 : "CASCADIA_HOSTING_WINDOW_CLASS"

```
So I'm curious.  How's that being done?  Does it _guarantee_ that the console window won't be seen?

**Reply by vefatica** [reply]
There's another artefact associated with the scenario of my last post.  If I dismiss TCC's dialog with a mouse click (OK/Cancel) then the busy mouse pointer (spinning blue ring) starts immediately (before the dialog disappears) and continues indefinitely after dialog disappears and the Cascadia window regains the foreground.  Even though it no longer seems necessary for my WinEvent hook to foreground the Cascadia window, I'm still using it to click (or wiggle) the mouse to get rid of the busy pointer.  I don't know why this happens in this particular scenario but IMHO, Windows is, in general, all too eager to show the busy mouse pointer.
